### PR TITLE
Added support for Parquet TIME

### LINKED ISSUE #4753: Support for Parquet DATE and TIME types
A customer needs us to support the Parquet [`DATE`](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#date) type. (A Java `java.time.LocalDate` column should be written as a Parquet `DATE`, and a Parquet `DATE` column should be read back into Java `LocalDate` columns.)

We might as well handle `java.time.LocalTime`/Parquet [`TIME`](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#time) while we're at it.

Relates to #3367.

### PR DESCRIPTION
**Pull request by malhotrashivam**
Closes #4753 

Interesting reference: https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#temporal-types

### CODE REVIEW

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/transfer/Int32CastablePrimitiveTransfer.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import io.deephaven.engine.table.ColumnSource;
+import org.jetbrains.annotations.NotNull;
+
+import java.nio.IntBuffer;
+
+/**
+ * A transfer object base class for primitive types that can be cast to {@code int} without loss of precision.
+ * Uses {@link IntBuffer} as the buffer type.
+ */
+// TODO Need a better name
```
Comment (malhotrashivam) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/transfer/Int32CastablePrimitiveTransfer.java:15]: Same as earlier

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/pagestore/topage/ToTimePage.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    @SuppressWarnings("rawtypes")
+    private static final ToPage MILLIS_INSTANCE = new ToTimePageFromMillis();
+    @SuppressWarnings("rawtypes")
+    private static final ToPage MICROS_INSTANCE = new ToTimePageFromMicros();
+    @SuppressWarnings("rawtypes")
+    private static final ToPage NANOS_INSTANCE = new ToTimePageFromNanos();
+
+    @SuppressWarnings("unchecked")
+    public static <ATTR extends Any> ToPage<ATTR, LocalTime[]> create(@NotNull final Class<?> nativeType,
+            final LogicalTypeAnnotation.TimeUnit unit, final boolean isAdjustedToUTC) {
```
Comment (rcaudy) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/pagestore/topage/ToTimePage.java:26]:
```suggestion
    public static <ATTR extends Any> ToPage<ATTR, LocalTime[]> create(
            @NotNull final Class<?> nativeType,
            @NotNull final LogicalTypeAnnotation.TimeUnit unit,
            @SuppressWarnings("unused") final boolean isAdjustedToUTC) {
```

File: py/server/tests/test_parquet.py [hunk 3]
```diff
@@ -354,10 +354,12 @@ def test_dictionary_encoding(self):
         self.assertTrue((metadata.row_group(0).column(2).path_in_schema == 'someIntColumn') &
                         ('RLE_DICTIONARY' not in str(metadata.row_group(0).column(2).encodings)))
 
-    def test_dates(self):
-        dh_table = empty_table(20).update(formulas=[
+    def test_dates_and_time(self):
```
Comment (rcaudy) [inline py/server/tests/test_parquet.py:357]: Are we able to force pyArrow or similar to write millis or micros for time, so we can verify reading for those?

File: replication/util/src/main/java/io/deephaven/replication/ReplicatePrimitiveCode.java [hunk 4]
```diff
@@ -465,8 +465,32 @@ private static void floatToAllFloatingPoints(String sourceClassJavaPath,
...
     public static String replaceAll(String sourceClassJavaPath, Map<String, Long> serialVersionUIDs,
             String[] exemptions, String[]... pairs) throws IOException {
+        final String resultClassJavaPath =
+                getResultClassPathFromSourceClassPath(sourceClassJavaPath, exemptions, pairs);
+        return replaceAll(sourceClassJavaPath, resultClassJavaPath, serialVersionUIDs, exemptions, pairs);
+    }
+
+    public static String replaceAll(String sourceClassJavaPath, String resultClassJavaPath,
+            Map<String, Long> serialVersionUIDs, String[] exemptions,
+            String[]... pairs) throws IOException {
```
Comment (rcaudy) [inline replication/util/src/main/java/io/deephaven/replication/ReplicatePrimitiveCode.java:485]:
```suggestion
            Map<String, Long> serialVersionUIDs, String[] exemptions, String[]... pairs) throws IOException {
```

File: replication/util/src/main/java/io/deephaven/replication/ReplicatePrimitiveCode.java [hunk 5]
```diff
@@ -475,17 +499,9 @@ public static String replaceAll(String sourceClassJavaPath, Map<String, Long> se
...
-
-        final String resultClassName = replaceAllInternal(sourceClassName, null, exemptions, pairs);
-        final String fullResultClassName = packageName + '.' + resultClassName;
-        final Long serialVersionUID = serialVersionUIDs == null ? null : serialVersionUIDs.get(fullResultClassName);
-        final String resultClassJavaPath = basePath(sourceClassJavaPath) + '/' + resultClassName + ".java";
-        final Map<String, List<String>> noReplicateParts = findNoLocateRegions(resultClassJavaPath);
-
         System.out.println("Generating java file " + resultClassJavaPath);
         String body = replaceAllInternal(inputText.toString(), serialVersionUID, exemptions, pairs);
+        final Map<String, List<String>> noReplicateParts = findNoLocateRegions(resultClassJavaPath);
```
Comment (rcaudy) [inline replication/util/src/main/java/io/deephaven/replication/ReplicatePrimitiveCode.java:504]: I assume you have run *all* of the replicators, including the reflective ones, and ensured that we get no unexpected changes?

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/transfer/PrimitiveTransferArrayBacked.java [hunk 6]
```diff
@@ -14,22 +14,25 @@
...
- * to a Buffer using {@link ColumnSource#fillChunk(ChunkSource.FillContext, WritableChunk, RowSequence)}.
+ * This is a generic class that can be used to transfer primitive data types directly from a {@link ColumnSource} to a
+ * {@link Buffer} using {@link ColumnSource#fillChunk(ChunkSource.FillContext, WritableChunk, RowSequence)}. This class
+ * can only be used if the {@link WritableChunk} and {@link Buffer} are backed by the same array.
  */
-abstract class PrimitiveTransfer<C extends WritableChunk<Values>, B extends Buffer> implements TransferObject<B> {
-    private final C chunk;
-    private final B buffer;
+// TODO Need a better name for this class
+abstract class PrimitiveTransferArrayBacked<CHUNK_TYPE extends WritableChunk<Values>, BUFFER_TYPE extends Buffer>
```
[the remark this answers is not in the document]
Comment (rcaudy) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/transfer/PrimitiveTransferArrayBacked.java:22]: Maybe `FillingPrimitiveTransfer` and `GettingPrimitiveTransfer`? The former uses `fillChunk`, the latter uses `getChunk`, right?

File: py/server/tests/test_parquet.py (hunk 3 above)
Comment (malhotrashivam) [reply py/server/tests/test_parquet.py:357]: Pyarrow uses micros by default, so was able to test that path.
For millis, I changed our writing code to write millis and then reran the tests. So couldn't add any tests in the pipeline for millis but tested it manually.
I can look into forcing it to use millis somehow to add tests for millis too.

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java [hunk 7]
```diff
@@ -2382,6 +2421,23 @@ public static long nanosOfDay(@Nullable final ZonedDateTime dateTime) {
         return epochNanos(dateTime) - epochNanos(atMidnight(dateTime));
     }
 
+    /**
+     * Returns the number of nanoseconds that have elapsed since the top of the day. The results will be adjusted as per
+     * the daylight savings time.
```
[the remark this answers is not in the document]
Comment (malhotrashivam) [reply engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:2426]: If you look at other `nanosOfDay` implementations, they don't adjust for daylight savings.
But this new one does adjust for daylight savings.
It becomes more clear when you look at the tests for these methods.

File: replication/util/src/main/java/io/deephaven/replication/ReplicatePrimitiveCode.java (hunk 5 above)
Comment (malhotrashivam) [reply replication/util/src/main/java/io/deephaven/replication/ReplicatePrimitiveCode.java:504]: Yup, tested (and now reran) all reflective and static replicators in replication/reflective and replication/static folders.

**Review by rcaudy — COMMENTED** [review]
.

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rcaudy) [reply engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java]: We confirmed that LocalTime.toNanoOfDay does no adjustment.

File: py/server/tests/test_parquet.py (hunk 3 above)
Comment (malhotrashivam) [reply py/server/tests/test_parquet.py:357]: Added a test for this

**Review by rcaudy — APPROVED** [review]

**Pull request merged**